[14.0][ADD] add new module resource_booking_auto_create

huguesdk opened the pull request:
and also add new modules resource_category and resource_category_booking.

this module provides a wizard that allows to select available resources by category to assign to a booking. resource combinations are handled automatically.

this module depends on #80.

huguesdk committed:
[ADD] new module resource_category

add new modules resource_category and resource_category_booking.

huguesdk committed:
[ADD] resource_booking_combination_auto_create

add new module resource_booking_combination_auto_create (in progress).

huguesdk committed:
[FIX] fix existing combination matching

huguesdk committed:
[FIX] disable resource combination auto-assignment

huguesdk committed:
[IMP] add test

huguesdk committed:
[IMP] select resource categories with checkboxes

select resource categories with checkboxes instead of editing a list
of resource categories with their quantity.

huguesdk committed:
[IMP] display resources on resource booking form

display the list of assigned resources instead of the resource
combination.

huguesdk committed:
[IMP] allow to unassign all resources

allow to unassign all resources from a booking.

huguesdk committed:
[IMP] display error if booking dates are not set

huguesdk committed:
[FIX] make resource_category_ids a computed field

change resource_category_ids from a regular many2many field (using the table
of resource.booking.category.selection as relation table) into a computed
field. the previous setup prevented the module to be installed (although it
worked like a charm on a database where the module was already installed).

note that this change introduces a bug that occurs when going back to the
first step of the wizard, unselecting a resource category and going to the
next step.

huguesdk committed:
[FIX] several fixes and improvements

* fix bug that appeared when going back to the first step, unchecking
  the first checked category and going to the next step. it was caused
  by an erroneous ondelete="cascade".
* change user message when no resources are available for the current
  resource category.
* ensure resource categories, configuration steps and resources are
  presented in alphabetical order.
* don't list resources as available if already selected in another step.
* keep selections across wizard steps.

huguesdk committed:
[FIX] don't list already booked resources

huguesdk committed:
[FIX] correct readme

huguesdk committed:
[IMP] restore selection when re-opening wizard

* restore selection when re-opening wizard (as much as possible: the
  previous selection cannot be exactly determined if selected resources
  are in multiple categories).
* raise validation error when no dates are set as soon as the wizard is
  opened instead of on step 2.
* fix availability of resources currently assigned to the booking and
  selected in other steps.

huguesdk committed:
[FIX] pass pre-commit

pedrobaeza commented:
Isn't better to add the category concept directly to the resource.booking model (so you can assign combinations or categories), and then adapt the selection code to take that into account?

carmenbianca commented:
@pedrobaeza I think it was written this way to stay compatible with the way of working in `resource_booking`. If the whole thing could be re-engineered from scratch (or if breaking compatibility is not a problem), it would probably be easier to remove the concept of 'combinations' `resource_booking`, and instead simply directly store the `resource.resource`s in `resource.booking`.

Subsequently, you could create modules that introduce the concept of combinations or categories to `resource.booking`.

Ripping out categories from `resource.booking` and replacing them with categories without doing the above refactoring seems like a recipe for bugs to me.

pedrobaeza commented:
What I'm thinking is complementary: just a new selector at resource.booking level to select also categories, and take into account all the resources of the combinations or the categories. They are minimal changes.

But as you prefer...

carmenbianca commented on resource_booking_combination_auto_create/models/resource_booking.py:
```diff
@@ -0,0 +1,36 @@
+from odoo import models
+
+
+class ResourceBooking(models.Model):
+    _inherit = "resource.booking"
+    # override constraint to allow to save resource bookings without
+    # combinations in pending state.
+    _sql_constraints = [
+        (
+            "combination_required_if_event",
+            "check(meeting_id is null or state = 'pending' or combination_id is not null)",
+            "Missing resource booking combination.",
+        ),
+    ]
```
This should be in upstream `resource_booking` probably.

huguesdk replied:
@pedrobaeza what do you think? this change is to allow to save a booking with a date but no resource combination, and leave it in `pending` state. currently this is not allowed by `resource_booking`, because as soon as a date is set, it considers the booking to be in `scheduled` state, and that requires a resource combination.

carmenbianca commented on resource_booking_combination_auto_create/models/resource_booking.py:
```diff
@@ -0,0 +1,36 @@
+        ),
+    ]
+
+    def _check_scheduling(self):
+        has_meeting = self.filtered("meeting_id")
+        if not has_meeting:
+            return
+        has_rbc = self.filtered("combination_id.resource_ids")
+        ready_rbc = has_meeting & has_rbc
+        # only check scheduling for bookings that have a meeting and a
+        # combination.
+        return super(ResourceBooking, ready_rbc)._check_scheduling()
```
idem

carmenbianca commented on resource_booking_combination_auto_create/models/resource_booking.py:
```diff
@@ -0,0 +1,36 @@
+        # combination.
+        return super(ResourceBooking, ready_rbc)._check_scheduling()
+
+    def _compute_state(self):
+        for rb in self:
+            super(ResourceBooking, rb)._compute_state()
+            if rb.state == "scheduled" and not rb.combination_id:
+                # stay in pending state when a start date is set but no
+                # combination.
+                rb.state = "pending"
```
idem

carmenbianca commented on resource_booking_combination_auto_create/tests/test_resource_combination_wizard.py:
```diff
@@ -0,0 +1,188 @@
+        displays an error message, and allows only to go back or cancel.
+        """
+
+    @freeze_time("2023-01-31 12:00:00")
```
I think you can use the freeze_time decorator on the class instead, to save yourself some effort.

huguesdk replied:
done in a later commit.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_category_selection.py:
```diff
@@ -0,0 +1,29 @@
+        required=True,
+        ondelete="cascade",
+    )
+    quantity = fields.Integer("Quantity", required=True, default=1)
```
This needs a comment or a help field.

huguesdk replied:
at first, the user would have to select a quantity of resources for each category, but this has been removed in a later commit.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+        "resource.booking.category.selection",
+        "Current Resource Category Selection",
+        ondelete="cascade",
+    )
```
A comment to describe what 'current' means is welcome here.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+    current_resource_booking_category_selection_id_id = fields.Integer(
+        "Current Resource Category Selection ID",
+        related="current_resource_booking_category_selection_id.id",
+    )
```
A comment to describe the necessity of this field would be nice.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+    current_resource_category_name = fields.Char(
+        related="current_resource_booking_category_selection_id"
+        ".resource_category_id.name",
+    )
```
idem

huguesdk replied:
this field has been removed in a later commit. it was used in the wizard to display the resource category of the current configuration step, but this is now handled by the `current_configuration_step_message` field.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+        "Number of Configuration Steps",
+        compute="_compute_configure_step_count",
+        store=True,
+    )
```
I think I would prefer 'total' or 'amount' over 'count' here. I will leave this to you.

huguesdk replied:
i didn’t like that name either. renamed to `num_configuration_steps`. what do you think?

carmenbianca replied:
Perfect!

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+        compute="_compute_selected_resource_ids",
+    )
+
+    @api.depends("resource_booking_category_selection_ids")
+    def _compute_configure_step_count(self):
+        self.configure_step_count = len(self.resource_booking_category_selection_ids)
```
Isn't there a for-loop over `self` missing here?

Idem for all other compute methods.

Is this because wizards don't generally get grouped together in a recordset?

huguesdk replied:
you’re right. i’ll fix all compute methods.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_category_selection.py:
```diff
@@ -0,0 +1,29 @@
+    resource_ids = fields.Many2many(
+        "resource.booking.category.selection.resource",
+        string="Resources",
+    )
```
To avoid confusion, this should be renamed to `selection_resource_ids`.

I found this rather confusing in the following bit of code:

```python
    @api.depends("resource_booking_category_selection_ids.resource_ids.resource_id")
    def _compute_selected_resource_ids(self):
        self.selected_resource_ids = self.mapped(
            "resource_booking_category_selection_ids.resource_ids.resource_id"
        )
```

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+    def _reopen_self(self):
+        act_window = super()._reopen_self()
+        act_window["name"] = self._compute_step_name()
+        return act_window
```
A comment describing what bug this fixes. I assume it's a workaround for something.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+    def _update_configure_step(self):
+        self.current_resource_booking_category_selection_id = (
+            self.resource_booking_category_selection_ids[self.configure_step]
+        )
```
Is there a reason this isn't computed?

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+            self.state = "final"
+        else:
+            self.configure_step += 1
+            self._update_configure_step()
```
Comments :) I think what I'm missing is, for each method, why/when the method might be called. The subsequent code seems fairly self-evident.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+        # todo: implement this
+        return True
+
+    def find_available_resources(self):
```
Method name lies. It finds-and-links available resources. I propose to either change the method name (something like 'update'), or to split into two methods if possible.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+                    }
+                )
+
+    def create_combination(self):
```
Making this as a note without yet having read the contents: I'm very surprised that this takes no arguments.

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -0,0 +1,209 @@
+        for resource_combination in existing_combinations:
+            # use first matching combination
+            if len(resource_combination.resource_ids) == len(resources):
+                break
```
Does this do anything?

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -182,9 +182,10 @@ def create_combination(self):
         resource_combination_model = self.env["resource.booking.combination"]
         existing_combinations = resource_combination_model.search(domain)
         resource_combination = resource_combination_model.browse()
-        for resource_combination in existing_combinations:
+        for combination in existing_combinations:
             # use first matching combination
-            if len(resource_combination.resource_ids) == len(resources):
+            if len(combination.resource_ids) == len(resources):
+                resource_combination = combination
                 break
```
Can't you use `(combination.resource_ids & resources) == resources`?

huguesdk replied:
the domain already ensures that the combination contains all the needed resources, but it can contain more. this is why comparing the length is enough and slightly more efficient.

carmenbianca replied:
ha, that's really cool

carmenbianca commented on resource_booking_combination_auto_create/wizards/resource_booking_combination_wizard.py:
```diff
@@ -187,11 +187,16 @@ def find_available_resources(self):
 
     def create_combination(self):
         # search for an existing combination that matched the selected resources.
+        resource_combination_model = self.env["resource.booking.combination"]
         resources = self.selected_resource_ids
+        if not resources:
+            self.resource_booking_id.combination_id = (
+                resource_combination_model.browse()
+            )
+            return
```
Why not set to False?

carmenbianca commented on resource_booking_combination_auto_create/tests/test_resource_combination_wizard.py:
```diff
@@ -301,6 +302,69 @@ def test_select_resource_in_multiple_selected_categories(self):
+            self._create_wizard_with_selected_categories(
+                booking, [self.room_category, self.worker_category]
+            )
+        self.assertEqual(
+            str(cm.exception),
+            "To select resources, the booking must have a start date and a "
+            "duration.",
+        )
```
Did cm survive the context? Shouldn't this be indented to the right?

huguesdk replied:
this seems very weird indeed, but that code would not be reachable if it would be indented to the right, as the context is exited as soon as the exception is raised. do you see another way of checking the message of the exception?

carmenbianca replied:
from a little searching on the internet; apparently this is completely normal. see also other example.

```python
🐚 coopiteasy reuse-tool main $ with open("README.md") as fp:
...............................     text = fp.read()
...............................
🐚 coopiteasy reuse-tool main $ fp
<_io.TextIOWrapper name='README.md' mode='r' encoding='UTF-8'>
🐚 coopiteasy reuse-tool main $ fp.closed
True
```

carmenbianca reviewed (changes requested):
Went through this commit-by-commit. It was a lot to go through; my level of scrutiny decreased during later commits.

Some comments may be made obsolete by later commits.

huguesdk committed:
[IMP] improve code after review

* rename wizard models to shorter and more understandable names (and
  rename fields accordingly).
* fix compute methods to correctly handle multiple records.
* make current_wizard_category_id (previously
  current_resource_booking_category_selection_id) a computed field and
  remove ._update_configure_step() method.
* rename "configure" to "configuration".
* rename configure_step_count field to num_configuration_steps.
* rename resource_ids field of rbc.wizard.category (previously
  resource.booking.category.selection) to selected_resource_ids.
* add comments.

huguesdk commented:
thanks a lot @carmenbianca for this in-depth review. all the changes you suggested are now made. i also renamed the wizard models to shorter and more understandable names.

houssine78 reviewed (approved):
LGTM
